### PR TITLE
add _TZE204_rzrrjkz2 Smart Sprinkler Timer

### PR DESCRIPTION
**Pull request by OgV1**
add _TZE204_rzrrjkz2 Smart Sprinkler Timer

### CODE REVIEW

File: src/devices/tuya.ts [hunk 1]
```diff
@@ -8734,6 +8734,45 @@ const definitions: Definition[] = [
...
+        fingerprint: tuya.fingerprint('TS0601', ['_TZE204_rzrrjkz2']),
+        model: 'TS0601',
+        vendor: 'Tuya',
+        fromZigbee: [tuya.fz.datapoints],
+        toZigbee: [tuya.tz.datapoints],
+        description: 'Smart Sprinkler Timer',
+        onEvent: tuya.onEventSetTime,
+        configure: tuya.configureMagicPacket,
+        exposes: [
+            e.switch(),
```
Comment (Koenkk) [inline src/devices/tuya.ts:8747]: I'm missing a lot of exposes here, could you add them?

File: src/devices/tuya.ts [hunk 2]
```diff
@@ -8734,6 +8734,45 @@ const definitions: Definition[] = [
...
+            tuya.exposes.batteryState(),
+            e.switch(),
+        ],
+        meta: {
+            tuyaDatapoints: [
+                [1, 'switch', tuya.valueConverter.onOff],
```
Comment (Koenkk) [inline src/devices/tuya.ts:8753]:
```suggestion
                [1, 'state', tuya.valueConverter.onOff],
```

File: src/devices/tuya.ts [hunk 3]
```diff
@@ -8734,6 +8734,45 @@ const definitions: Definition[] = [
             ],
         },
     },
+    {
+        fingerprint: tuya.fingerprint('TS0601', ['_TZE204_rzrrjkz2']),
+        model: 'TS0601',
```
Comment (Koenkk) [inline src/devices/tuya.ts:8739]: Please use the model number listed on the device or product site here.

File: src/devices/tuya.ts (hunk 1 above)
Comment (OgV1) [reply src/devices/tuya.ts:8747]: Hi Koenkk, 

Thank you for the fast respond, really appreciated,

I must say that I'm really new to GitHub and actually have a little knowledge on how to make exposes works. 
The data point IDs are extracted from the device in Tuya Cloud Project, as I'm following from this instruction
https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html#_3-understanding-tuya-datapoints

I tried to put a lot of exposes before but it seems to be rejected, that's why I have to remove them.

Could you please provide me any instructions on how to make exposes works?

I can see these information on Device Debugging on Tuya Cloud Project, Not sure if it should help us on developing exposes?

<html>
<body>

Code | Type | Values
-- | -- | --
switch | Boolean | "{true,false}"
status | Enum | {   "range": [     "off",     "auto",     "disabled"   ] }
countdown | Integer | {   "unit": "min",   "min": 1,   "max": 240,   "scale": 0,   "step": 1 }
countdown_left | Integer | {   "unit": "min",   "min": 1,   "max": 240,   "scale": 0,   "step": 1 }
water_current | Integer | {   "unit": "gal/min",   "min": 0,   "max": 1000000000,   "scale": 3,   "step": 1 }
battery_percentage | Integer | {   "unit": "%",   "min": 0,   "max": 100,   "scale": 0,   "step": 1 }
water_total | Integer | {   "unit": "gal",   "min": 0,   "max": 999999999,   "scale": 3,   "step": 1 }
weather_delay | Enum | {   "range": [     "24h",     "48h",     "72h",     "cancel"   ] }
normal_timer | String | {   "maxlen": 255 }
weather_switch | Boolean | "{true,false}"
switch_enabled | Boolean | "{true,false}"


</body>
</html>

Comment (Koenkk) [reply src/devices/tuya.ts:8747]: > I tried to put a lot of exposes before but it seems to be rejected, that's why I have to remove them.

What do you mean with "rejected" here? 

For examples, please look at e.g. this:

expose
https://github.com/Koenkk/zigbee-herdsman-converters/blob/4812c1bc77683beb786644de86ac8c642d895a50/src/devices/tuya.ts#L1104

datapoint:
https://github.com/Koenkk/zigbee-herdsman-converters/blob/4812c1bc77683beb786644de86ac8c642d895a50/src/devices/tuya.ts#L1127

File: src/devices/tuya.ts [hunk 4]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
             ],
         },
     },
+    {
+        fingerprint: tuya.fingerprint('NAS-WV03B', ['_TZE204_rzrrjkz2']),
+        model: 'NAS-WV03B',
+        vendor: 'NEO',
+        fromZigbee: [tuya.fz.datapoints],
+        toZigbee: [tuya.tz.datapoints],
+        description: 'Smart Sprinkler Timer',
```
Comment (Koenkk) [inline src/devices/tuya.ts:8743]:
```suggestion
        description: 'Smart sprinkler timer',
```

File: src/devices/tuya.ts [hunk 5]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
             ],
         },
     },
+    {
```
Comment (Koenkk) [inline src/devices/tuya.ts:8737]: Please move this definition to `neo.ts`

File: src/devices/tuya.ts [hunk 6]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
...
+        fingerprint: tuya.fingerprint('NAS-WV03B', ['_TZE204_rzrrjkz2']),
```
Comment (Koenkk) [inline src/devices/tuya.ts:8738]:
```suggestion
        zigbeeModel: ['NAS-WV03B'],
```

The modelId is unique, so no need to use fingerprint

File: src/devices/tuya.ts [hunk 7]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
...
+        model: 'NAS-WV03B',
+        vendor: 'NEO',
+        fromZigbee: [tuya.fz.datapoints],
+        toZigbee: [tuya.tz.datapoints],
+        description: 'Smart Sprinkler Timer',
+        onEvent: tuya.onEventSetTime,
+        configure: tuya.configureMagicPacket,
+        exposes: [
+            e.switch().withEndpoint('switch').withDescription('Switch'),
```
Comment (Koenkk) [inline src/devices/tuya.ts:8747]:
```suggestion
            e.switch(),
```

Does this work?

File: src/devices/tuya.ts [hunk 8]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
...
+            e.enum('status', ea.STATE, ['off', 'auto', 'disabled']).withDescription('Status'),
+            e.numeric('countdown', ea.STATE_SET).withUnit('min').withValueMin(1).withValueMax(240).withDescription('Countdown'),
+            e.numeric('countdown_left', ea.STATE).withUnit('min').withValueMin(1).withValueMax(240).withDescription('Countdown Left'),
```
Comment (Koenkk) [inline src/devices/tuya.ts:8750]:
```suggestion
            e.numeric('countdown_left', ea.STATE).withUnit('min').withValueMin(1).withValueMax(240).withDescription('Countdown left'),
```

Comment (Koenkk) [reply src/devices/tuya.ts:8750]: Also make all the other descriptions lower case (except the first letter)

File: src/devices/tuya.ts [hunk 9]
```diff
@@ -8734,6 +8734,66 @@ const definitions: Definition[] = [
...
+            e.numeric('battery_percentage', ea.STATE).withUnit('%').withValueMin(0).withValueMax(100).withDescription('Battery Percentage'),
+            e.numeric('water_total', ea.STATE).withUnit('L').withValueMin(0).withValueMax(378541.0).withValueStep(0.001)
+                .withDescription('Total Water Flow (L)'),
+            e.enum('weather_delay', ea.STATE_SET, ['24h', '48h', '72h', 'cancel']).withDescription('Weather Delay'),
+            e.text('normal_timer', ea.STATE_SET).withDescription('Normal Timer'),
+            e.binary('weather_switch', ea.STATE_SET, 'ON', 'OFF').withDescription('Weather Switch'),
+            e.binary('switch_enabled', ea.STATE_SET, 'ON', 'OFF').withDescription('Switch Enabled'),
+            e.numeric('surplus_flow', ea.STATE).withDescription('Surplus Flow'),
+            e.numeric('single_watering_duration', ea.STATE).withDescription('Single Watering Duration'),
+            e.numeric('interface_refresh', ea.STATE).withDescription('Interface Refresh'),
```
Comment (Koenkk) [inline src/devices/tuya.ts:8762]: Could you improve the descriptions? E.g. for this one it's quite unclear what it does.

File: src/devices/tuya.ts (hunk 5 above)
Comment (OgV1) [reply src/devices/tuya.ts:8737]: Do I need to create a new pull for this?

**Comment by OgV1** [comment]
some how it shows

> Duplicate zigbee model NAS-WV03B
> 
>       272 |                 device.zigbeeModel.forEach((m) => {
>       273 |                     if (foundZigbeeModels.includes(m.toLowerCase())) {
>     > 274 |                         throw new Error(`Duplicate zigbee model ${m}`)

and I'm not sure, what I do wrong?

File: src/devices/tuya.ts (hunk 9 above)
Comment (OgV1) [reply src/devices/tuya.ts:8762]: Done

File: src/devices/tuya.ts (hunk 8 above)
Comment (OgV1) [reply src/devices/tuya.ts:8750]: Done

**Comment by OgV1** [comment]
Wait, I think I figured out :)

**Comment by Koenkk** [comment]
Thanks!

Could you also submit a picture for the docs? This can be done by clicking [here](https://github.com/Koenkk/zigbee2mqtt.io/upload/master/public/images/devices). 

Make sure that:
- The filename is `MODEL.png` (update `model` accordingly) 
- The size is 512x512
- The background is transparent (use e.g. [Adobe remove background](https://new.express.adobe.com/tools/remove-background))

**Pull request merged**